### PR TITLE
Adds support for Gzip extension in CSV exporter

### PR DESCRIPTION
**Pull request by athityakumar**
Closes issue #8

### CODE REVIEW

File: lib/daru/io/exporters/csv.rb [hunk 1]
```diff
@@ -37,30 +37,45 @@ class CSV < Base
...
-          writer << @dataframe.vectors.to_a unless @headers == false
+          contents = process_dataframe
 
-          @dataframe.each_row do |row|
-            writer << if @convert_comma
-                        row.map { |v| v.to_s.tr('.', ',') }
-                      else
-                        row.to_a
-                      end
+          if @compression == :gzip
```
Comment (zverok) [inline lib/daru/io/exporters/csv.rb]: I believe that the most important part of @pathtm's suggestion in https://github.com/SciRuby/daru/issues/127#issuecomment-313414356 is _default_ value of compression option equal to `:infer` (e.g. "quess by file extension"). Any other option (e.g. `compression: :algo`) is necessary only for complicated cases like "gzipped data directly received from/sent to network as a string".

File: lib/daru/io/exporters/csv.rb [hunk 2]
```diff
@@ -37,30 +37,45 @@ class CSV < Base
...
+            ::Zlib::GzipWriter.open(@path) do |gz|
+              contents.each { |content| gz.write("#{content.join(',')}\n") }
```
Comment (zverok) [inline lib/daru/io/exporters/csv.rb]: I believe it is too naive. CSV is not only about commas (though, even commas can be rewritten with options, in Russian and Ukrainian locales typically comma is decimal separator, therefore typical CSV separator is semicolon), but also about properly handling string fields (including multiline strings), dates and so on. So, compression should be somehow "pipelined" to file creation, not replace original CSV creation code.

File: lib/daru/io/exporters/csv.rb [hunk 3]
```diff
@@ -37,30 +37,45 @@ class CSV < Base
...
+          order = [@dataframe.vectors.to_a] unless @headers == false
+          data  = @dataframe.map_rows do |row|
+            next row.to_a unless @convert_comma
+            row.map { |v| v.to_s.tr('.', ',') }
           end
 
-          writer.close
+          return data  if order.nil?
+          return order if data.empty?
+          order + data
```
Comment (zverok) [inline lib/daru/io/exporters/csv.rb]:
```ruby
[].tap do |result|
  result << [@dataframe.vectors.to_a] unless @headers == false
  result.concat(@dataframe.map_rows { ... })
end
```

File: lib/daru/io/exporters/csv.rb [hunk 4]
```diff
@@ -37,30 +37,45 @@ class CSV < Base
...
+          end
+        end
+
+        private
+
+        def process_dataframe
+          order = [@dataframe.vectors.to_a] unless @headers == false
+          data  = @dataframe.map_rows do |row|
+            next row.to_a unless @convert_comma
+            row.map { |v| v.to_s.tr('.', ',') }
```
Comment (zverok) [inline lib/daru/io/exporters/csv.rb]: This code would behave REALLY funny in dataframe looking like:
```
John Doe|200.00
Jane Foe|300.50
Jim Dowson Sr.|183.20
George T. Sorrensen|1240.00
```
...don't you think?..

File: lib/daru/io/exporters/csv.rb (hunk 1 above)
Comment (v0dro) [inline lib/daru/io/exporters/csv.rb]: Hmmm it makes sense to default to `:infer` and then use appropriate algorithm based on file extension.

File: lib/daru/io/exporters/csv.rb (hunk 4 above)
Comment (v0dro) [inline lib/daru/io/exporters/csv.rb]: @athityakumar doesn't the Ruby standard library have a CSV writer that you can use out of the box? I'm sure they must have handled these problems already.

File: lib/daru/io/exporters/csv.rb (hunk 2 above)
Comment (v0dro) [inline lib/daru/io/exporters/csv.rb]: I second that.

File: lib/daru/io/exporters/csv.rb [hunk 5]
```diff
@@ -37,30 +40,48 @@ class CSV < Base
...
+        def compression?(algorithm, *formats)
+          @compression == algorithm || formats.any? { |f| @path.end_with?(f) }
+        end
+
+        def process_dataframe
+          [].tap do |result|
+            result << @dataframe.vectors.to_a unless @headers == false
+            @dataframe.map_rows do |row|
+              next result << row.to_a unless @convert_comma
+              result << row.map { |v| v.to_s.tr('.', ',') }
```
Comment (zverok) [inline lib/daru/io/exporters/csv.rb]: I believe we discussed the problem with this approach elsewhere? Or is it fixed in another PR?..

Comment (athityakumar) [inline lib/daru/io/exporters/csv.rb]: There was no option in CSV stdlib to change the decimal point to comma. Fixed this with commit 8538770 - `:convert_comma` now affects only float values and not strings that contain dots.

**Pull request merged**